crates/sui-indexer-alt-consistent-store/src/restore/mod.rs

```diff
@@ -76,28 +75,19 @@ pub(crate) struct Restorer<S: Schema> {
     /// Channels to send live object partitions down, one per pipeline being restored.
     restore_tx: BTreeMap<String, mpsc::Sender<Arc<LiveObjects>>>,
 
-    /// Handles for tasks spawned by the restorer. They return `Ok(_)` if they complete
-    /// successfully, or `Err(_)` otherwise (they encountered an error, or they were cancelled).
-    handles: Vec<JoinHandle<Result<(), ()>>>,
+    /// Services spawned by the restorer, for individual pipelines.
+    workers: Vec<Service>,
 
     /// Number of object files to download concurrenctly
     object_file_concurrency: usize,
 
     /// Maximum size of the backlog of object files waiting to be processed by one worker.
     object_file_buffer_size: usize,
-
-    /// Cancellation token shared among all continuous tasks in the service.
-    cancel: CancellationToken,
 }
 
-/// Internal type used by tasks to propagate errors or shutdown signals.
-#[derive(thiserror::Error, Debug)]
-enum Break {
-    #[error("Shutdown received")]
-    Cancel,
-
-    #[error(transparent)]
-    Err(#[from] anyhow::Error),
+pub struct Finalizer {
```

Comment:
Not entirely happy with this -- the restore task needs to do some clean-up work at the end assuming the main service exits completely cleanly. I implemented this by having the `run` task output a `Service` for the main service, and then this `Finalizer` which can be run as its own service, when the main service completes successfully.

Originally, I tried to add finalization to the `Service` abstraction, but it doesn't compose well -- in particular with finalization usually you want to say things like "once this set of tasks completes successfully, run this task", but the way we merge services together can cause a finalizer task to end up waiting arbitrarily long after the tasks it cares about have completed to run, which is usually not desired.